Add different trust levels to AppStore interface

LukasReschke opened the pull request:
:construction: :construction: :construction:

TODO:
- [x] Have apps.owncloud.com return the trust levels – somehow (cc @karlitschek)
- [x] Make it look nicer (cc @jancborchardt)
- [x] Move the "Experimental applications ahead" bar right before the experimental ones 
- [x] Show the experimental applications at the end of the site if 
- [x] Make the "Show experimental applications" button actually working

:construction: :construction: :construction:

Heads-up: We don't know when an application is activated (i.e. already installing) whether it is an official one / approved one or complete random stuff. Thus in the "activated" and "not activated" view we only show the official apps where we know it from the shipped flag and everything else has no flag applied.

Current layout:

![2015-03-30_15-52-45](https://cloud.githubusercontent.com/assets/878997/6897806/30c19272-d6f6-11e4-9716-84e4506a48b4.png)
![2015-03-30_15-54-51](https://cloud.githubusercontent.com/assets/878997/6897812/363f1404-d6f6-11e4-97d2-45624f5e2200.png)
![2015-03-30_15-56-41](https://cloud.githubusercontent.com/assets/878997/6897815/3acce83e-d6f6-11e4-923c-2de159bab4af.png)

LukasReschke committed:
Add experimental applications switch

Allows administrators to disable or enabled experimental applications as well as show the trust level.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/66750efe-976c-4779-abce-2def4d1692b1) was created.

LukasReschke commented:
Interesting. Seems like there is a SimpleXML version on our CI server that behaves differently than mine :disappointed: 

```
:22:09 1) OCSClientTest::testGetCategoriesParseError
10:22:09 simplexml_load_string(): Entity: line 1: parser error : Start tag expected, '<' not found
10:22:09 
10:22:09 /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/lib/private/ocsclient.php:114
10:22:09 /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/tests/lib/OCSClientTest.php:178
10:22:09 
10:22:09 2) OCSClientTest::testGetApplicationsParseError
10:22:09 simplexml_load_string(): Entity: line 1: parser error : Start tag expected, '<' not found
10:22:09 
10:22:09 /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/lib/private/ocsclient.php:178
10:22:09 /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/tests/lib/OCSClientTest.php:374
10:22:09 
10:22:09 3) OCSClientTest::testGetApplicationParseError
10:22:09 simplexml_load_string(): Entity: line 1: parser error : Start tag expected, '<' not found
10:22:09 
10:22:09 /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/lib/private/ocsclient.php:259
10:22:09 /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/tests/lib/OCSClientTest.php:636
10:22:09 
10:22:09 4) OCSClientTest::testGetApplicationDownloadParseError
10:22:09 simplexml_load_string(): Entity: line 1: parser error : Start tag expected, '<' not found
10:22:09 
10:22:09 /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/lib/private/ocsclient.php:323
10:22:09 /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/tests/lib/OCSClientTest.php:867
```

… digging …

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/d1cb5025-112f-45e2-9efe-688921640763) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/66750efe-976c-4779-abce-2def4d1692b1) completed: **101 new issues, 21 updated code elements**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/d1cb5025-112f-45e2-9efe-688921640763) completed: **No issues found**

jancborchardt commented on settings/templates/apps.php:
```diff
@@ -95,8 +98,35 @@
+				'enabling experimental applications you may experience data ' .
+				'loss or install potential insecure or malicious software.')) ?>
+		</p>
+	</div>
```
Instead of a permanent header, can you insert a `<div class="section">` with that info between the last »Approved« and the first »Experimental« app?

LukasReschke replied:
Guess what it says in the top post:
>  Move the "Experimental applications ahead" bar right before the experimental ones

jancborchardt commented on settings/js/apps.js:
```diff
@@ -9,6 +9,17 @@ Handlebars.registerHelper('score', function() {
+			return new Handlebars.SafeString('<span class="experimental">Experimental</span>');
+		}
+	}
+});
```
Instead of returning a span with the text here, can you give the `<div class="app-level">` the class `official` etc and then only add the text inside? That makes it much easier to style.

jospoortvliet commented:
@LukasReschke could you use a different wording to describe the experimental apps? I'd like this:
[] show experimental apps
Experimental apps are not checked for security issues and are new or known to be unstable and under heavy development. Installing these can cause data loss or security breaches.

Use the same in the yellow header, saying:
"display of experimental apps enabled"
You might want to include a 'hide experimental apps' button there, as well as a 'dismiss warning' one.

jancborchardt commented on tests/lib/OCSClientTest.php:
```diff
@@ -0,0 +1,928 @@
+				   <id>168707</id>
+				   <name>Calendar 8.0</name>
+				   <version>0.6.4</version>
+				   <label>recommended</label>
```
So now we can remove the »Recommended« label, right?

LukasReschke replied:
Already is removed in the web interface. That one is in the response that the server returns. It's an unit test. Probably we need to keep it on the server as well for longer or it won't work at all for older versions.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/e2800cfe-8397-4ce2-b246-db6f9ab8396e) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/e2800cfe-8397-4ce2-b246-db6f9ab8396e) completed: **66 new issues, 24 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/330ea7b8-67d1-4803-8d99-375230c28abc) was created.

LukasReschke commented:
@karlitschek @jancborchardt @jospoortvliet Please take a look - should work now - not the best code but I added at least some unit tests for other untested stuff :see_no_evil: 

Please test properly, in my case it works good. To test this checkout this branch and – this is important – change the `$OC_VERSION` in `version.php` to `array(8,0,0,1)` so that the AppStore will show you the 8.0 apps. If you then switch to a section such as "Tool" you should see some experimental applications.

The yellow banner might require some CSS tweaks - but I'm not a design guy and for me everything looks good as long as it has a color ;-) – I also added the recommendations by @jospoortvliet in the text field as well as the tipsy overlay on the labels.

Also: Please ensure to clean your browser cache before testing this.

Some small remarks that we can't address at the moment:

1. The "Activated" view will only show tags for the "Official" apps, the thing is we really don't know at this step which state an application has as we are not connected  to the appstore and addressing this properly should be done in a refactoring that I wanted to avoid.
2. Clicking the "Enable experimental apps" button will reload the page and you have to go to the category you were before manually. That is caused by the reason that we don't recognize the current category by the URL and is also there in master. Should be addressed in the future but is out-of-scope for this PR.

And yes, some code of this is quite hacky but I couldn't come up with something better in this short timeframe since I have to do other stuff as well… But we always can improve in the future anyway :-)

LukasReschke commented:
Some screenshots:

![2015-04-02_12-47-31](https://cloud.githubusercontent.com/assets/878997/6962636/85eb5a70-d936-11e4-81c7-57082c2e565c.png)
![2015-04-02_12-47-36](https://cloud.githubusercontent.com/assets/878997/6962638/8792a9fa-d936-11e4-9025-7ddcc7bd50ec.png)
![2015-04-02_12-47-41](https://cloud.githubusercontent.com/assets/878997/6962639/8acae218-d936-11e4-9a26-4c9066d86045.png)
![2015-04-02_12-47-51](https://cloud.githubusercontent.com/assets/878997/6962640/8cb60300-d936-11e4-9e8b-3c14f258f446.png)

LukasReschke commented:
I think we can do other improvements such as showing the author etc. in other PRs – these changes are actually much easier (about 1 line additional PHP and then the usual CSS /HTML thing) but require that somehow (*hint hint* @jancborchardt  *hint hint*) comes up with a design for those ;-)

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/330ea7b8-67d1-4803-8d99-375230c28abc) completed: **86 new issues, 36 updated code elements**

jospoortvliet commented:
I'd make the popup text this: "This app is not checked for security issues and are new or known to be unstable. Install on your own risk."

Look-wise I have no comments (I'm sure Jan does). Functionality-wise, I agree that this, once working and safe (not breaking things) should go in before we iterate and improve. As a first step I'm happy with this, even if we don't get to improve things before the 8.1 release.

jospoortvliet commented:
That is a first :+1: provided the popup text is improved but note I didn't (and won't) test.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/c223c4b9-fe99-44d4-a63a-824cb7728d52) was created.

LukasReschke commented:
Adjusted. Thanks @jospoortvliet.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/c223c4b9-fe99-44d4-a63a-824cb7728d52) completed: **86 new issues, 36 updated code elements**

karlitschek commented:
Awesome!!! I would add a "It is not recommended to install this app on a production server" to the warning. But beside that really good :+1:

karlitschek commented:
And I agree that a link to the author page would be nice :-)

LukasReschke committed:
Make user names clickable

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/b7571aa8-1ea8-49c7-b43f-6600d4d12313) was created.

LukasReschke commented:
Rebased. The usernames are now also clickable. I use the `profilepage` entry for this. However, this means it links for example to http://opendesktop.org/usermanager/search.php?username=arkascha instead of https://apps.owncloud.com/usermanager/search.php?username=arkascha. 

@karlitschek Would it be easier to add a hack to core or change this in the appstore code?

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/6ecdd75f-c849-49e7-a1ed-dcf3d3dbae03) was created.

LukasReschke commented:
@karlitschek Also this means we have to ensure that `profilepage` will never contain user controllable input, or at least check it begins with `http://` or `https://` otherwise if somebody inserts a `profilepage` like `javascript:alert(1)` this could be potentially bad…

LukasReschke commented:
Mhm. Maybe I can use a JS or PHP function to ensure this in the ownCloud code base as well.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/b7571aa8-1ea8-49c7-b43f-6600d4d12313) completed: **86 new issues, 36 updated code elements**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/6ecdd75f-c849-49e7-a1ed-dcf3d3dbae03) completed: **86 new issues, 36 updated code elements**

jancborchardt committed:
app mgmt app level design enhancements

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/e6a762b7-1e79-4cc4-9121-eb1dc7b11cfa) was created.

jancborchardt commented:
@LukasReschke added some design enhancements to the tags.

I can’t fully test it though since for the pages for market apps I get »no apps for your version«. Why that?

LukasReschke commented:
> I can’t fully test it though since for the pages for market apps I get »no apps for your version«. Why that?

Changed the `$OC_Version` to 8.0? i.e. `$OC_Version=array(8, 0, 0, 1);` in `version.php` ? :-)

jancborchardt commented:
Oh and btw, shouldn’t every app get a badge? Even those downloaded via zip or git.

LukasReschke commented:
> Oh and btw, shouldn’t every app get a badge? Even those downloaded via zip or git.

See initial post from me:
> Heads-up: We don't know when an application is activated (i.e. already installing) whether it is an official one / approved one or complete random stuff. Thus in the "activated" and "not activated" view we only show the official apps where we know it from the shipped flag and everything else has no flag applied.

We just don't know that at this state and adding such blackmagic looks really out-of-scope for me for this PR as it requires quite some other hacks / refactoring. Something for the future.

LukasReschke commented:
![2015-04-07_12-00-19](https://cloud.githubusercontent.com/assets/878997/7020999/b3fe5846-dd1d-11e4-9c7f-c8c7f2b7c997.png)
![2015-04-07_12-00-29](https://cloud.githubusercontent.com/assets/878997/7021000/b582b41e-dd1d-11e4-9694-51ee7a8dcabf.png)

jancborchardt committed:
apps mgmt: style experimental apps section, small design fixes

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/ced2f273-37e1-4e38-964b-d6ea9dfa8b34) was created.

jancborchardt commented:
Pushed more design fixes, done now from my view. :)

* styled »experimental apps« section properly
* moved star rating to right of badge so badge is always to the left, and fixed alignment
* unrelated but annoying: capitalized »Other« category title in sidebar

jancborchardt commented:
@owncloud/designers a review please. :)

LukasReschke commented:
@owncloud-bot Retest this please.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/e6a762b7-1e79-4cc4-9121-eb1dc7b11cfa) completed: **84 new issues, 36 updated code elements**

MorrisJobke commented:
Looks good :)

LukasReschke commented:
@karlitschek @DeepDiver1975 Do you guys already have done the fight whether 8.1 or 8.2? :see_no_evil:

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/ced2f273-37e1-4e38-964b-d6ea9dfa8b34) completed: **84 new issues, 36 updated code elements**

DeepDiver1975 commented:
> @karlitschek @DeepDiver1975 Do you guys already have done the fight whether 8.1 or 8.2? :see_no_evil:

no

DeepDiver1975 commented:
> The inspection completed: 84 new issues, 36 updated code elements

@LukasReschke mind taking a look? THX

LukasReschke committed:
Deduplicate code

LukasReschke commented:
> @LukasReschke mind taking a look? THX

74 of those are in the test files due to me being unable to properly typehint stuff… Though we do it that way all over the code. If somebody has suggestions how to solve this: Let me know :-)

1 is a duplication of the constructor in appsettingscontroller.php, not really solvable.

1 is a "the parameter $s is not used and could be removed" in server.php – I'm not sure if that has any side-effects? – It is there everywhere else as well.

1 doc comment patch is in code that got not even touched, I adjusted these anyways.

3 are a duplication in `lib/private/ocsclient.php` – these could indeed be improved – I pushed something that should help here.

LukasReschke committed:
Fix PHPDoc

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/64799459-5d52-4ec9-8bee-16fdf99204cc) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/49232d83-6e42-47e8-9217-53b538c29b75) was created.

LukasReschke commented:
@owncloud-bot Retest this please.

karlitschek commented:
ok. as discussed with @DeepDiver1975 and @LukasReschke we should put this into 8.1

MorrisJobke commented:
Code looks good and it works :+1:

PVince81 commented:
"Install on your own risk" => "Install at your own risk" http://ell.stackexchange.com/questions/13918/at-your-own-risk-or-on-your-own-risk

PVince81 commented:
@Xenopathic @tomneedham native language checks please ?

I guess this still has time to get translated right ?

PVince81 commented on config/config.sample.php:
```diff
@@ -577,6 +577,15 @@
 'appstoreurl' => 'https://api.owncloud.com/v1',
 
 /**
+ * Whether to experimental apps in the appstore interface
```
"to experimental" => "to show" ? Or does it remove the checkbox ?

Does the checkbox change that setting ?

PVince81 replied:
Tested this, this controls the checkbox state.

Xenopathic replied:
This should probably be 'Whether to *show* experimental apps...'

LukasReschke replied:
Indeed :See

PVince81 commented on settings/js/apps.js:
```diff
@@ -438,6 +463,16 @@ OC.Settings.Apps = OC.Settings.Apps || {
 		});
 
+		$(document).on('click', '#enable-experimental-apps', function () {
+			var state = $('#enable-experimental-apps').prop('checked');
```
Here you could use `$(this).prop('checked')`, where `this` is the clicked object.

PVince81 commented:
Tested, works.

Can you add JS unit tests for testing the sort algo ?
There is already test code that renders each app block one by one, will need to be changed to call a mocked loadCategory with possibly a mocked response. Let me know if you need help.

LukasReschke commented:
Let me look into that tomorrow. Glad that you didn't find anything other grave :smile:

Xenopathic commented on settings/js/apps.js:
```diff
@@ -81,13 +91,27 @@ OC.Settings.Apps = OC.Settings.Apps || {
 					$('#apps-list-empty').removeClass('hidden');
 				}
+
+				$('.app-level .official').tipsy({fallback: t('core', 'Official apps are developed by and within the ownCloud community and its GitHub repository and offer functionality central to ownCloud. They are ready for serious use.')});
```
Too many 'and's in this sentence, you might want to reword it and/or add some commas. Also, 'serious use' => 'production use'

LukasReschke committed:
Fix wording

LukasReschke committed:
Reference `$(this)` instead

LukasReschke committed:
Add JS unit test for sorting

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/a3b70aee-07c8-450a-ba8c-c28a8b81782e) was created.

LukasReschke commented:
Changed the wording and added a JS unit test for sorting as well.

LukasReschke commented:
Not now @owncloud-bot …

LukasReschke commented:
@owncloud-bot Retest this please.

MorrisJobke commented:
My :+1: is still valid

PVince81 commented:
:+1:

LukasReschke commented:
Awesome. Thanks for the review @PVince81 @MorrisJobke. And kudos for the nice design to @jancborchardt :-)

LukasReschke commented:
@karlitschek Just to clarify some things about apps.owncloud.com, from my understanding:

1. When the user that created the app is the "owncloud" user it will get marked official?
2. When I mark an app as approved it will get marked as approved?
3. When "maximum" required =< the current ownCloud version and "minimum required" >= the current ownCloud version it will get shown in the ownCloud interface ?

Are these assumptions correct?

karlitschek commented:
1. No. Being marked as "official" requires that an admin set´s this flag. I think this is the correct behavior. What do you think?
2. yes
3. yes

LukasReschke commented:
> No. Being marked as "official" requires that an admin set´s this flag. I think this is the correct behavior. What do you think?

That makes sense. So just to re-verify, if I'm modifying an existing app as admin:

- Approval
  - Approved: will get marked as approved  
  - Not approved: Experimental
- Category
  - Recommended: ??
  - 3rdparty: ??

Guess I will only concentrate on the "Approval" switch then ,_)

jancborchardt commented:
@karlitschek @LukasReschke since the new Official/Approved/Experimental is the new system, we can remove the Recommended/3rdparty, right?

karlitschek commented:
@jancborchardt Yes. I would replace it with the new terms

LukasReschke commented:
> @karlitschek @LukasReschke since the new Official/Approved/Experimental is the new system, we can remove the Recommended/3rdparty, right?

Well the terms are already gone in the current web interface as far I can tell – I referred to the usage on apps.owncloud.com :-)

MorrisJobke commented:
> No. Being marked as "official" requires that an admin set´s this flag. I think this is the correct behavior. What do you think?
> 
> That makes sense. So just to re-verify, if I'm modifying an existing app as admin:
> 
>     Approval
>         Approved: will get marked as approved
>         Not approved: Experimental

Can we document this somewhere? Because there will be people that ask for these terms and their meaning.

jancborchardt commented:
Yes – basically there should just be one selection in apps.owncloud.com called »App trust level«. Default to »Experimental« for new apps, and app store reviewers can set it to »Approved«. (And only admins could set apps to »Official«.)
